lib/ansible/modules/files/archive.py

```diff
@@ -211,133 +211,152 @@ def main():
         # If we actually matched multiple files or TRIED to, then
         # treat this as a multi-file archive
-        archive = globby or os.path.isdir(expanded_paths[0]) or len(expanded_paths) > 1
+        archive = globby or os.path.isdir(b_expanded_paths[0]) or len(b_expanded_paths) > 1
     else:
         archive = True
 
     # Default created file name (for single-file archives) to
     # <file>.<format>
-    if not dest and not archive:
-        dest = '%s.%s' % (expanded_paths[0], format)
+    if not b_dest and not archive:
+        b_dest = b'%s.%s' % (b_expanded_paths[0], b_fmt)
 
     # Force archives to specify 'dest'
-    if archive and not dest:
+    if archive and not b_dest:
         module.fail_json(dest=dest, path=', '.join(paths), msg='Error, must specify "dest" when archiving multiple files or trees')
 
-    archive_paths = []
-    missing = []
-    arcroot = ''
+    b_sep = to_bytes(os.sep, errors='surrogate_or_strict')
 
-    for path in expanded_paths:
+    b_archive_paths = []
+    b_missing = []
+    b_arcroot = b''
+
+    for b_path in b_expanded_paths:
         # Use the longest common directory name among all the files
         # as the archive root path
-        if arcroot == '':
-            arcroot = os.path.dirname(path) + os.sep
+        if b_arcroot == b'':
+            b_arcroot = os.path.dirname(b_path) + b_sep
         else:
-            for i in range(len(arcroot)):
-                if path[i] != arcroot[i]:
+            for i in range(len(b_arcroot)):
+                if b_path[i] != b_arcroot[i]:
                     break
 
-            if i < len(arcroot):
-                arcroot = os.path.dirname(arcroot[0:i + 1])
+            if i < len(b_arcroot):
+                b_arcroot = os.path.dirname(b_arcroot[0:i + 1])
 
-            arcroot += os.sep
+            b_arcroot += b_sep
 
         # Don't allow archives to be created anywhere within paths to be removed
-        if remove and os.path.isdir(path):
-            path_dir = path
-            if path[-1] != '/':
-                path_dir += '/'
-
-            if dest.startswith(path_dir):
-                module.fail_json(path=', '.join(paths), msg='Error, created archive can not be contained in source paths when remove=True')
-
-        if os.path.lexists(path) and path not in expanded_exclude_paths:
-            archive_paths.append(path)
+        if remove and os.path.isdir(b_path):
+            b_path_dir = b_path
+            if not b_path.endswith(b'/'):
+                b_path_dir += b'/'
+
+            if b_dest.startswith(b_path_dir):
+                module.fail_json(
+                    path=', '.join(paths),
+                    msg='Error, created archive can not be contained in source paths when remove=True'
+                )
+
+        if os.path.lexists(b_path) and b_path not in b_expanded_exclude_paths:
+            b_archive_paths.append(b_path)
         else:
-            missing.append(path)
+            b_missing.append(path)
```

Comment:
Seems like this should either be ```b_missing.append(b_path)``` or ```missing.append(path)```.  I don't see missing being compared or combined with any other strings later, so it may be that the latter is fine.  Then again, it may be confusing to have a single variable which contains non-byte paths so it might be more readable to use the former.

Reply:
Yep, that's my mistake. It should have been `b_path`. We're currently looping `b_expanded_paths`, so I wanted to just stick with bytes at this point.